diff --git a/build.gradle b/build.gradle
--- a/build.gradle
+++ b/build.gradle
@@ -104,42 +104,17 @@
 //}
 
 dependencies {
-    implementation 'com.github.albfernandez:juniversalchardet:2.5.0'
     implementation files("../tools/batmass-io-1.36.5.jar")
-    implementation 'com.google.code.gson:gson:2.12.1'
-    implementation 'one.util:streamex:0.8.3'
-    implementation 'org.jooq:jool-java-8:0.9.15'
-    implementation 'org.apache.maven:maven-artifact:3.9.9'
-    implementation 'org.jgrapht:jgrapht-core:1.5.2'
-    implementation 'com.github.vlsi.mxgraph:jgraphx:4.2.2'
 
-    implementation group: 'org.apache.commons', name: 'commons-compress', version: '1.27.1'
 
-    implementation 'org.apache.commons:commons-lang3:3.17.0'
-    implementation 'com.miglayout:miglayout-swing:11.4.2'
-    implementation 'org.apache.commons:commons-text:1.13.0'
 
-    implementation 'io.reactivex.rxjava3:rxjava:3.1.10'
 
-    implementation 'org.greenrobot:eventbus-java:3.3.1'
-    implementation 'org.greenrobot:essentials:3.1.0'
-    implementation 'org.jsoup:jsoup:1.19.1'
     implementation 'com.squareup.okio:okio:3.10.2'
-        implementation 'com.squareup.okhttp3:okhttp:4.12.0'
 
-    implementation 'net.java.balloontip:balloontip:1.2.4.1'
-    implementation group: 'org.netbeans.api', name: 'org-netbeans-swing-outline', version: 'RELEASE250'
-    implementation 'commons-io:commons-io:2.18.0'
 
-    implementation 'ch.qos.logback:logback-classic:1.5.18'
-    implementation 'ch.qos.logback:logback-core:1.5.18'
 
-    implementation 'org.apache.parquet:parquet-hadoop:1.15.1'
-    implementation 'org.apache.hadoop:hadoop-common:3.4.1'
-    implementation 'org.apache.hadoop:hadoop-mapreduce-client-core:3.4.1'
 
     // Use JUnit test framework
-    testImplementation 'junit:junit:4.13.2'
 }
 
 test {
