## FragPipe workflows

Listed below are the analysis workflows provided with FragPipe. Workflows can be customized and saved (with a unique name) for later use. Visit [workflow sharing](https://github.com/Nesvilab/FragPipe/tree/gh-pages/docs/workflow-sharing) to browse additional workflows or upload your own to share with other FragPipe users.


#### Closed (standard) database searches
* [Default](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#default)

#### DIA
* [DIA_SpecLib_Quant](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#dia_speclib_quant)
* [DIA_SpecLib_Quant_Phospho](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#dia_speclib_quant_phospho)
* [DIA_SpecLib_Quant_diaPASEF](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#dia_speclib_quant_diapasef)
* [DIA_SpecLib_Quant_Phospho_diaPASEF](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#dia_speclib_quant_phospho_diapasef)
* [DIA_DIA-Umpire_SpecLib_Quant](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#dia_dia-umpire_speclib_quant)

#### Non-specific digestion
* [Nonspecific-HLA](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#nonspecific-hla)
* [Nonspecific-HLA-C57](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#nonspecific-hla-c57)
* [Nonspecific-HLA-TMT10](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#nonspecific-hla-tmt10)
* [Nonspecific-HLA-phospho](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#nonspecific-hla-phospho)
* [Nonspecific-HLA-glyco](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#nonspecific-hla-glyco)
* [Nonspecific-HLA-DIA](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#nonspecific-hla-dia)
* [Nonspecific-HLA-DIA-Astral](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#nonspecific-hla-dia-astral)
* [Nonspecific-HLA-diaPASEF](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#nonspecific-hla-diapasef)
* [Nonspecific-peptidome](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#nonspecific-peptidome)

#### DDA label-free quantification
* [LFQ-MBR](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#lfq-mbr)
* [LFQ-phospho](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#lfq-phospho)
* [LFQ-ubiquitin](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#lfq-ubiquitin)

#### Isobaric label-based quantification
* [TMT10](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt10)
* [TMT10-MS3](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt10-ms3)
* [TMT10-bridge](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt10-bridge)
* [TMT10-phospho](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt10-phospho)
* [TMT10-MS3-phospho](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt10-ms3-phospho)
* [TMT10-phospho-bridge](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt10-phospho-bridge)
* [TMT10-acetyl](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt10-acetyl)
* [TMT10-acetyl-noloc](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt10-acetyl-noloc)
* [TMT10-Open](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt10-open)
* [TMT10-ubiquitin](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt10-ubiquitin)
* [TMT10-ubiquitination-K_tmt_or_ubiq](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt10-ubiquitination-k_tmt_or_ubiq)
* [TMT10-ubiquitination-K_tmt_plus_ubiq](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt10-ubiquitination-k_tmt_plus_ubiq)
* [TMT16](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt16)
* [TMT16-MS3](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt16-ms3)
* [TMT16-phospho](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt16-phospho)
* [TMT16-acetyl](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt16-acetyl)
* [TMT16-acetyl-noloc](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt16-acetyl-noloc)
* [TMT16-ubiquitination-K_tmt_or_ubiq](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt16-ubiquitination-k_tmt_or_ubiq)
* [TMT16-ubiquitination-K_tmt_plus_ubiq](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#tmt16-ubiquitination-k_tmt_plus_ubiq)
* [iTRAQ4](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#itraq4)
* [iTRAQ4-phospho](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#itraq4-phospho)

#### MS1 label-based quantification
* [SILAC3](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#silac3)
* [SILAC3-phospho](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#silac3-phospho)

#### Open and mass offset
* [Open](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#open)
* [Mass-Offset-CommonPTMs](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#mass-offset-commonptms)
* [Open-quickscan](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#open-quickscan)
* [Diagnostic-ion-mining](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#diagnostic-ion-mining)
* [XRNAX-MassOffset](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#xrnax-massoffset)
* [FPOP](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#fpop)

#### Labile PTM Searches
* [Labile_phospho](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#labile_phospho)
* [Labile_ADP-ribosylation](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#labile_adp-ribosylation)

#### Glyco
* [glyco-N-HCD](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-n-hcd)
* [glyco-N-Hybrid](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-n-hybrid)
* [glyco-N-LFQ](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-n-lfq)
* [glyco-N-open-HCD](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-n-open-hcd)
* [glyco-N-open-Hybrid](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-n-open-hybrid)
* [glyco-N-TMT](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-n-tmt)
* [glyco-O-HCD](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-o-hcd)
* [glyco-O-Hybrid](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-o-hybrid)
* [glyco-O-open-HCD](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-o-open-hcd)
* [glyco-O-open-Hybrid](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-o-open-hybrid)
* [glyco-O-Pair](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-o-pair)
* [glyco-N-DIA](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-n-dia)
* [glyco-O-DIA-HCD](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-o-dia-hcd)
* [glyco-O-DIA-OPair](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#glyco-o-dia-opair)

#### Chemoproteomics
* [isoDTB-ABPP](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#isodtb-abpp)
* [diaTOP_ABPP](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#diatop_abpp)
* [ipIAA-ABPP](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#ipiaa-abpp)
* [isoTOP-ABPP](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#isotop-abpp)
* [SLC-ABPP](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#slc-abpp)
* [PAL](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#pal)

#### Wide-window DDA
* [WWA](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe_workflows.html#wwa)

<br>

##### Default
Simple closed search, no quantification. MSFragger search with 'stricttrypsin' (Trypsin/P) enzyme, fully tryptic peptides only, up to 2 missed cleavages. Met oxidation and protein N-term Acetyl specified as variable modifications, and C+57 as fixed modification. Deisotoping, mass calibration, and parameter optimization are enabled. Post-processing with Philosopher (PeptideProphet, ProteinProphet), with 1% FDR filtering at the PSM and protein levels (sequential filtering).
<br>

##### DIA_SpecLib_Quant
Complete workflow for DIA analysis using spectral library building and quantification using DIA-NN. Spectral library can be built from DDA data and/or DIA data, with direct identification from DIA data using MSFragger-DIA. All data are processed together using MSFragger (in DDA or DIA mode, respectively) - MSBooster - Percolator - ProteinProphet (Philosopher) - EasyPQP. The library is filtered to 1% FDR at the protein and peptide levels. DIA files annotated as 'DIA-Quant' are used for quantification only (i.e. not used for spectral library building). If building the library from fractionated DDA data only, in EasyPQP tab change RT Calibration option to "ciRT", or "iRT" if using organisms other than yeast or human (requires iRT peptides). Supports DDA RAW/.d files, mzML, and MGF files. DIA files with staggered windows should be mzML files demultiplexed using Proteowizard. For quantification using DIA-NN, Thermo/Sciex DIA files should be in mzML format, and .d for Bruker's timsTOF. diaPASEF DIA data will only be used for quantification (i.e. automatically considered as 'DIA-Quant' data type).
<br>

##### DIA_SpecLib_Quant_Phospho

<br>

##### DIA_SpecLib_Quant_diaPASEF

<br>

##### DIA_SpecLib_Quant_Phospho_diaPASEF

<br>

##### DIA_DIA-Umpire_SpecLib_Quant
Alternative workflow for DIA analysis using spectral library building and quantification using DIA-NN. This is an alternative to the main DIA_SpecLib_Quant workflow in that DIA data (files annotated as \u2018DIA\u2019 type) are first converted to pseudo-MS/MS spectra using DIA-Umpire, followed by conventional MSFragger (DDA mode) search.Spectral library can be built from DIA (via DIA-Umpire) and optionally DDA data. All data (DDA and pseudo-MS/MS spectra from DIA) are processed together using MSFragger - MSBooster - Percolator - ProteinProphet (Philosopher) - EasyPQP. The library is filtered to 1% FDR at the protein and peptide levels. Supports Thermo and Sciex data only. For spectral library building both RAW and mzML files are supported, however, quantification with DIA-NN requires mzML files.
<br>

##### Nonspecific-HLA
Nonspecific search, with recommended settings for HLA peptides. Peptide length 7-25. MSFragger search assumes cysteines were not alkylated (i.e. samples were not treated with iodoacetamide). Cysteinylation (C+119) is specified as variable modification. MSFragger rescoring with MSBooster. PSM validation with Percolator. Protein FDR filter is not applied, so each output file (PSM, ion, peptide) is filtered to 1% FDR at that level. If needed, extend to add label-free quantification (using IonQuant) or spectral library building with EasyPQP.
<br>

##### Nonspecific-HLA-C57
Nonspecific search, with recommended settings for HLA peptides. Peptide length 7-25. MSFragger search assumes cysteines were alkylated. Cysteinylation minus 57 (C+62, 62+57=119) is specified as variable modification. MSFragger rescoring with MSBooster. PSM validation with Percolator. Protein FDR filter is not applied, so each output file (PSM, ion, peptide) is filtered to 1% FDR at that level. If needed, extend to add label-free quantification (using IonQuant) or spectral library building with EasyPQP.
<br>

##### Nonspecific-HLA-TMT10
Wokflow for TMT10 labeled HLA peptidome data. Nonspecific search. MSFragger search assumes cysteines were not alkylated (i.e. samples were not treated with iodoacetamide). Cysteinylation (C+119) is specified as variable modification. TMT10 on peptide n-term and K as fixed modifications. MSFragger rescoring with MSBooster. PSM validation with Percolator. Protein FDR filter is not applied, so each output file (PSM, ion, peptide) is filtered to 1% FDR at that level. TMT quantification with TMT-Integrator.
<br>

##### Nonspecific-HLA-phospho
Workflow for identification of phosphopeptides in HLA peptidome data. Nonspecific search, peptide length 7-20. MSFragger search assumes cysteines were not alkylated (i.e. samples were not treated with iodoacetamide). Optionally add C+119 as variable mod. PSM rescoring and validation with MSBooster and Percolator. Group-specific FDR filtering (unmodified, peptides with common mods, and phosphopeptides are filtered separately). Protein FDR filter of 1%. Optionally add LFQ with IonQuant.
<br>

##### Nonspecific-HLA-glyco
Workflow for identification of glycopeptides in HLA peptidome data. Nonspecific search, peptide length 7-25, N-linked glyco mode settings (198 glycan list). MSFragger search assumes cysteines were not alkylated (i.e. samples were not treated with iodoacetamide). Optionally add C+119 as variable mod. PSM validation with PeptideProphet (glyco mode settings).Class specific FDR filtering (unmodified, peptides with common mods, and glycopeptides are filtered separately). Protein FDR filter of 1%. PTM-Shepherd glycan assignment. Optionally add LFQ with FreeQuant.
<br>

##### Nonspecific-HLA-DIA
Workflow for quantitative HLA peptidome profiling using DIA. Supports DDA-based (when DDA data is available), DIA-based (directDIA via DIA-Umpire) and hybrid (from both DIA and DDA) spectral library building. DIA files are first processed using DIA-Umpire. Pseudo-MS/MS spectra extracted by DIA-Umpire, and DDA data (when available), are then processed together using MSFragger nonspecific search, with recommended settings for HLA peptides. Peptide length 7-25, cysteines assumed to be not alkylated (i.e. samples were not treated with iodoacetamide). Cysteinylation (C+119) is specified as a variable modification. Deep learning-based rescoring of MSFragger results with MSBooster, followed by PSM validation with Percolator. Protein inference is performed with ProteinProphet but protein FDR filter is not applied, so each output file (PSM, ion, peptide) is filtered to 1% FDR at that level. Spectral library building with EasyPQP followed by quantification, using that library, from the DIA files using DIA-NN.
<br>

##### Nonspecific-HLA-DIA-Astral
Workflow for quantitative HLA type I peptidome profiling using Astral narrow window DIA. Supports DDA-based (when DDA data is available), direct Astral DIA-based, and hybrid (from both DIA and DDA) spectral library building. Narrow window DIA files are searched directly using MSFragger-DIA. Nonspecific search, peptide length 7-15, cysteines assumed to be not alkylated (i.e. samples were not treated with iodoacetamide). Cysteinylation (C+119) is specified as a variable modification. Deep learning-based rescoring of MSFragger results with MSBooster, followed by PSM validation with Percolator. Protein inference is performed with ProteinProphet but protein FDR filter is not applied, so each output file (PSM, ion, peptide) is filtered to 1% FDR at that level. Spectral library building with EasyPQP (the library is filtered to 1% peptide-level FDR) followed by quantification, using that library, from the DIA files using DIA-NN. Notes: for HLA type II peptides, increase max peptide length to 20 or 25. If samples were alkylated, specify C+57 as fixed and C+61.98264 as variable modification (Cysteinylation).
<br>

##### Nonspecific-HLA-diaPASEF
Workflow for quantitative HLA peptidome profiling using diaPASEF. Supports DIA-based (directDIA via diaTracer) and hybrid (from both DIA and DDA) spectral library building. DIA files are first processed using diaTracer. Pseudo-MS/MS spectra extracted by diaTracer, and DDA data (when available), are then processed together using MSFragger nonspecific search, with recommended settings for HLA peptides. Peptide length 7-25, cysteines assumed to be not alkylated (i.e. samples were not treated with iodoacetamide). Cysteinylation (C+119) is specified as a variable modification. Deep learning-based rescoring of MSFragger results with MSBooster, followed by PSM validation with Percolator. Protein inference is performed with ProteinProphet but protein FDR filter is not applied, so each output file (PSM, ion, peptide) is filtered to 1% FDR at that level. Spectral library building with EasyPQP followed by quantification, using that library, from the DIA files using DIA-NN.
<br>

##### Nonspecific-peptidome
Nonspecific search, with recommended settings for peptidome data (plasma, CSF, etc.). Peptide length 7-65. MSFragger search assumes cysteines were alkylated. Met oxidation, C-term amidantion, and Pyro-Glu are specified as variable modifications. Protein FDR filter is not applied, so each output file (PSM, ion, peptide) is filtered to 1% FDR at that level.
<br>

##### LFQ-MBR
Perform closed search, followed by label free quantification with match-between-runs using IonQuant. If using mzML files, need to choose the right MS data type (Regular MS vs IM-MS). Need to assign runs to experiments.
<br>

##### LFQ-phospho
Default phosphoproteomics workflow with label-free quantification by IonQuant. Runs should be annotated to experiments (enable MBR if using multiple experiments). PSM validation with MSbooster and Percolator, protein inference with ProteinProphet, site localization with PTM-Prophet.
<br>

##### LFQ-ubiquitin
LFQ workflow for ubiquitin enriched data, with quantification from MS1 with IonQuant. Oxidation and lysine ubiquitination are specified as variable modifications. Up to 3 missed cleavages. PSM validation with Percolator, and protein inference with ProteinProphet. Data is filtered at 1% FDR at the PSM, ion, peptide, and protein levels. Site localization with PTM-Prophet. Quantification with IonQuant.
<br>

##### TMT10
Basic TMT 10-plex workflow, with identification and quantification from high mass accuracy MS2. Met oxidation, protein N-term Acetyl, n-term TMT, and TMT on S ("overlabeling") are specified as variable modifications. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene level. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.
<br>

##### TMT10-MS3
TMT 10-plex workflow, with quantification from MS3 and identification from low mass accuracy MS2. Met oxidation, protein N-term Acetyl, and n-term TMT are specified as variable modifications. PSM validation with Percolator. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene level. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.
<br>

##### TMT10-bridge
TMT 10-plex, quantification and identification from high mass accuracy MS2. Met oxidation, protein N-term Acetyl, n-term TMT, and TMT on S ("overlabeling") are specified as variable modifications. TMT-Integrator with Bridge channel (labeled as 'pool' in the annotation files), data summarization at the gene level. Printing results with three normalization options (None; MD: Median Centering; GN: median centering with MAD variance scaling.
<br>

##### TMT10-phospho
TMT 10-plex workflow for phosphopeptide enriched data, with quantification from MS2. PTMProphet for site localization. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene/protein/peptide/site levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.
<br>

##### TMT10-phospho-bridge
TMT 10-plex workflow for phosphopeptide enriched data, with quantification from MS2. PTMProphet for site localization. TMT-Integrator with Bridge channel (labeled as 'pool' in the annotation files), median-centering normalization, data summarization at the gene/protein/peptide/site levels.
<br>

##### TMT10-MS3-phospho
TMT 10-plex workflow for phosphopeptide enriched data, with quantification from MS3 and identification from low resolution MS2. PTMProphet for site localization. PSM validation with Percolator. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene/protein/peptide/site levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.
<br>

##### TMT10-acetyl
TMT 10-plex workflow for acetylation enriched data, with quantification from MS2. TMT is specified as fixed modification on K and as variable on n-term. Oxidation, acetylation, carbamylation, and deamidation are specified as variable modifications. Up to 4 missed cleavages and 4 variable modifications in total. PSM validation with Percolator, and protein inference with ProteinProphet. Data is filtered at 1% FDR at the PSM, ion, peptide, and protein levels. Site localization with PTMProphet. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene/protein/peptide/site levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.
<br>

##### TMT10-acetyl-noloc
TMT 10-plex workflow for acetylation enriched data, with quantification from MS2. Site localization based on the MSFragger search engine assignment. TMT is specified as fixed modification on n-term and K. Oxidation is specified as variable modifications. Acetyl is specified as -187 (mass of Acetyl minus TMT, as these two modifications are not expected at the same time on the same residue). Also allowed is Carbamyl, specified as -187 (mass of carbamyl minus TMT). Up to 4 missed cleavages and 4 variable modifications in total. PSM validation with Percolator, and protein inference with ProteinProphet. Data is filtered at 1% FDR at the PSM, ion, peptide, and protein levels. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene/protein/peptide/site levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.
<br>

##### TMT10-Open
Open search workflow for TMT10 data. MSFragger localization-aware open search (LOS) algorithm, with deisotoping, mass calibration, parameter optimization, and monoisotope correction enabled. Mass range -150 to 500 Da, with Met oxidation included as variable modifications and TMT on n-term and K as fixed modifications. PeptideProphet with extended mass model. Crystal-C for artifact removal. PTM-Shepherd for mass shift summarization. TMT-Integrator for TMT quantification.
<br>

##### TMT10-ubiquitin
TMT 10-plex workflow for ubiquitin enriched data, with quantification from MS2. Site localization based on the MSFragger search engine assignment. TMT is specified as fixed on K and n-term. Oxidation is specified as variable modification. Ubiq is specified as two variable modifications: +114 (Ubiq in addition to TMT on the same residue) and -115 (just Ubiq, without TMT on the same residue). Up to 3 missed cleavages and 4 variable modifications in total. PSM validation with Percolator, and protein inference with ProteinProphet. Data is filtered at 1% FDR at the PSM, ion, peptide, and protein levels. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene/protein/peptide/site levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.
<br>

##### TMT10-ubiquitination-K_tmt_or_ubiq
TMT 10-plex workflow for ubiquitination enriched data, with quantification from MS2. This workflow assumes that lysine cannot be ubiquitinated and labeled with TMT at the same time. TMT is specified as fixed mod n-term and as variable mod on K. Oxidation and Ubiq (+114 on K) are specified as variable modifications. Modification stacking is not allowed. Up to 3 missed cleavages and 4 variable modifications in total. PSM validation with Percolator, and protein inference with ProteinProphet. Site localization with PTMProphet. Data is filtered at 1% FDR at the PSM, ion, peptide, and protein levels. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene/protein/peptide/site levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.

##### TMT10-ubiquitination-K_tmt_plus_ubiq
TMT 10-plex workflow for ubiquitination enriched data, with quantification from MS2. This workflow assumes that the ubiquitinated lysine is also labeled with TMT (for a combined modification mass of +343). TMT is specified as fixed mod on K and n-term. Oxidation and Ubiq (+114 on K) are specified as variable modifications. Up to 3 missed cleavages and 4 variable modifications in total. PSM validation with Percolator, and protein inference with ProteinProphet. Site localization with PTMProphet. Data is filtered at 1% FDR at the PSM, ion, peptide, and protein levels. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene/protein/peptide/site levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.

##### TMT16
Basic TMT 16-plex workflow, with quantification and identification from MS2. Met oxidation, protein N-term Acetyl, n-term TMT are specified as variable modifications. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene level. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.
<br>

##### TMT16-MS3
Basic TMT 16-plex workflow, with quantification from MS3 and identification form low mass accuracy MS2 (ion trap). Met oxidation, protein N-term Acetyl, n-term TMT are specified as variable modifications. PSM validation with Percolator. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene level. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.
<br>

##### TMT16-phospho
TMT 16-plex workflow for phosphopeptide enriched data, with quantification and identification from MS2. Met oxidation and Phospho are specified as variable modifications, TMT as fixed. Phosphosite localization using PTM-Prophet. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at all levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.

##### TMT16-acetyl
TMT 16-plex workflow for acetylation enriched data, with quantification from MS2. Lysine acetylation site localization with PTM-Prophet. TMT is specified as fixed modification on n-term and as variable modification on K. Oxidation is specified as variable modifications. Acetyl (+42) and Carbamyl (+43) are specified as variable modifications on K. Up to 4 missed cleavages and 4 variable modifications in total. PSM validation with Percolator, and protein inference with ProteinProphet. Data is filtered at 1% FDR at the PSM, ion, peptide, and protein levels. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene/protein/peptide/site levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.

##### TMT16-acetyl-noloc
TMT 16-plex workflow for acetylation enriched data, with quantification from MS2. Site localization based on the MSFragger search engine assignment. TMT is specified as fixed modification on n-term and K. Oxidation is specified as variable modifications. Acetyl is specified as -262 (mass of Acetyl minus TMT, as these two modifications are not expected at the same time on the same residue). Also allowed is Carbamyl, specified as -261 (mass of carbamyl minus TMT). Up to 4 missed cleavages and 4 variable modifications in total. PSM validation with Percolator, and protein inference with ProteinProphet. Data is filtered at 1% FDR at the PSM, ion, peptide, and protein levels. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene/protein/peptide/site levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.

##### TMT16-ubiquitination-K_tmt_or_ubiq
TMT 16-plex workflow for ubiquitination enriched data, with quantification from MS2. This workflow assumes that lysine cannot be ubiquitinated and labeled with TMT at the same time. TMT is specified as fixed mod n-term and as variable mod on K. Oxidation and Ubiq (+114 on K) are specified as variable modifications. Modification stacking is not allowed. Up to 3 missed cleavages and 4 variable modifications in total. PSM validation with Percolator, and protein inference with ProteinProphet. Site localization with PTMProphet. Data is filtered at 1% FDR at the PSM, ion, peptide, and protein levels. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene/protein/peptide/site levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.

##### TMT16-ubiquitination-K_tmt_plus_ubiq
TMT 16-plex workflow for ubiquitination enriched data, with quantification from MS2. This workflow assumes that the ubiquitinated lysine is also labeled with TMT (for a combined modification mass of +418). TMT is specified as fixed mod on K and n-term. Oxidation and Ubiq (+114 on K) are specified as variable modifications. Up to 3 missed cleavages and 4 variable modifications in total. PSM validation with Percolator, and protein inference with ProteinProphet. Site localization with PTMProphet. Data is filtered at 1% FDR at the PSM, ion, peptide, and protein levels. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene/protein/peptide/site levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.

##### iTRAQ4
Basic iTRAQ 4-plex workflow, with identification and quantification from high mass accuracy MS2. Met oxidation, protein N-term Acetyl, n-term TMT, and TMT on S ("overlabeling") are specified as variable modifications. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene level. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.
<br>

##### iTRAQ4-phospho
iTRAQ 4-plex workflow for phosphopeptide enriched data, with quantification from MS2. PTMProphet for site localization. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at the gene/protein/peptide/site levels. If a reference/bridge sample is available, specify the corresponding channel/sample name tag in the annotation file(s) and in TMT-Integrator tab.
<br>

##### SILAC3
Triple-SILAC quantification workflow. Closed search with MSFragger, SILAC quantification with IonQuant. Note: If the dataset has multiple MS files (e.g. replicates or LC fractions) they should be annotated as different experiments (e.g. F1, F2, etc. in the Experiment column).
<br>

##### SILAC3-phospho
Triple-SILAC, phosphopeptide-enriched workflow. Closed search with MSFragger followed by deep learning rescoring with MSBooster and PSM validation with Percolator, protein inference with ProteinProphet, SILAC quantification with IonQuant. PTM-Prophet (optional) will add localization probabilities to PSM files. Note: If the dataset has multiple MS files (e.g. replicates or LC fractions) they should be annotated as different experiments (e.g. F1, F2, etc. in the Experiment column).
<br> 

##### Open
Open search workflow for PTM analysis. MSFragger localization-aware open search (LOS) algorithm, with deisotoping, mass calibration, parameter optimization, and monoisotope correction enabled. Mass range -150 to 500 Da. PeptideProphet with extended mass model. Crystal-C for artifact removal. Precursor ion quantification using Freequant. PTM-Shepherd for mass shift summarization. For faster run time, in MSFragger change the number of allowed missed cleavages to 1.
<br>

##### Mass-Offset-CommonPTMs
Mass Offset (also known as Multinotch) search workflow for a fast search for most common modifications (list of mass shifts specified in MSFragger 'Mass Offset' field). MSFragger localization-aware open search (LOS) algorithm, filtered to report PSMs with specified mass shifts only (with isotope errors allowed). No variable modifications are specified. Mass calibration, parameter optimization, and precursor monoisotope error correction are enabled. PSM validation using PeptideProphet with extended mass model. Protein inference with ProteinProphet. Data filtered using 1% protein level FDR and additionally 1% FDR at the peptide/ion/PSM level. Precursor ion quantification using Freequant. PTM-Shepherd for mass shift summarization.
<br>

##### Open-quickscan
"Quick scan" version of the Open search workflow for PTM analysis. Compared to the full ('Open') workflow, the localization-aware open search (LOS) option is turned off; parameter optimization is turned off. Mass range is reduced to -20 to 500 Da. The number of missed cleavages is reduced to 1. No variable modifications are specified. This workflow has been designed to enable quick open searches (e.g. for tutorial purposes), for users with very slow computers, or when only a rough idea of the modification landscape is needed.
<br>

##### Diagnostic-ion-mining
Open search workflow for PTM analysis with diagnostic ion mining. MSFragger localization-aware open search (LOS) algorithm, with deisotoping, mass calibration, parameter optimization, and monoisotope correction enabled. Calibrated, decharged, and deisotoped mgfs are written for PTM-Shepherd. Mass range -20 to 500 Da. PeptideProphet with extended mass model. Crystal-C for artifact removal. Precursor ion quantification using Freequant. PTM-Shepherd for mass shift summarization and diagnostic ion mining. For faster run time, in MSFragger change the number of allowed missed cleavages to 1.
<br>

##### XRNAX-MassOffset
Wofkflow for the analysis of XRNAX protein-RNA crosslinks data using Mass Offset search. As an alternative to full open search, this workflow is more sensitive for identification of expected mass shifts (common RNA adducts). For description of the XRNAX teprotocol see "The Human RNA-Binding Proteome and Its Dynamics during Translational Arrest", Trendel J, Schwarzl T, Horos R, Prakash A, Bateman A, Hentze MW, Krijgsveld J. Cell. 2019 176:391-403.e19. PMID: 30528433.
<br>

##### FPOP
Workflow for analysis of FPOP data using a hybrid (variable modification and detailed mass offset) search.Very narrow exclusion range. Group-based FDR filtering included.
<br>

##### Labile_phospho
Searches for phosphorylation with combined labile/nonlabile search. Uses accurate mass modeling in PeptideProphet and PTMProphet for localization. Settings are for HCD/CID data. For Hybrid (EThcD, etc), change MSFragger ion types and diagnostic fragment settings.
<br>

##### Labile_ADP-ribosylation
Searches for ADP-ribosylation with combined labile/nonlabile search. Settings are for HCD/CID data. For Hybrid (EThcD, etc), change MSFragger ion types and diagnostic fragment settings.
<br>

##### glyco-N-HCD
For CID/HCD search of enriched N-glycopeptides. Performs MSFragger glyco search and applies glycan assignment with FDR control in PTM-Shepherd.
<br>

##### glyco-N-Hybrid
For hybrid activation (EThcD, etc) search of enriched N-glycopeptides.
<br>

##### glyco-N-LFQ
For search and label-free quantitation of enriched N-glycopeptides fragmented with CID/HCD. Performs glycan assignment/FDR control in PTM-Shepherd prior to quant.
<br>

##### glyco-N-open-HCD
For CID/HCD open search of N-glycopeptides. NOTE: glycan assignment/FDR control is NOT yet supported for open searches and is not performed in this workflow.
<br>

##### glyco-N-open-Hybrid
For hybrid activation open search of N-glycopeptides. NOTE: glycan assignment/FDR control is NOT yet supported for open searches and is not performed in this workflow.
<br>

##### glyco-N-TMT
For search and TMT quantitation of enriched N-glycopeptides fragmented with CID/HCD. Performs MSFragger glyco search, glycan FDR control in PTM-Shepherd, and TMT quant/summarization with TMT-Integrator. Settings are provided for TMT-10 with virtual reference channel - method can be adapted for other TMT settings by adjusting TMT-Integrator parameters.
<br>

##### glyco-O-HCD
For CID/HCD search of enriched O-glycopeptides.
<br>

##### glyco-O-Hybrid
For hybrid activation (EThcD, etc) search of enriched O-glycopeptides. MSFragger search for b,y,c,z,Y ions, O-Pair localization (using same scan for search and localization).
<br>

##### glyco-O-open-HCD
For CID/HCD open search of O-glycopeptides. NOTE: glycan assignment/FDR control is NOT yet supported for open searches and is not performed in this workflow.
<br>

##### glyco-O-open-Hybrid
For hybrid activation open search of O-glycopeptides. NOTE\: glycan assignment/FDR control is NOT yet supported for open searches and is not performed in this workflow.
<br>

##### glyco-O-Pair
For MSFragger search and O-Pair localization of O-glycopeptides. Requires paired scan data\! (for example, HCD-pd-EThcD).
<br>

##### glyco-N-DIA
Runs MSFragger-Glyco and PTM-Shepherd to build a spectral library from DDA n-glyco data, then quantifies DIA data using the library. Uses glycan FDR filtering from PTM-Shepherd when building the library. Load both DDA and DIA data on the workflow tab and set the Data Type to "DIA-quant" for all DIA files.
<br>

##### glyco-O-DIA-HCD
Runs MSFragger-Glyco to build a spectral library from HCD DDA data, then quantifies DIA data using the library. Note: library is NOT site-specific for HCD-only DDA data, glycans will be placed on the first allowed site in the peptide. Load both DDA and DIA data on the workflow tab and set the Data Type to "DIA-quant" for all DIA files.
<br>

##### glyco-O-DIA-OPair
Runs MSFragger-Glyco and O-Pair to build a spectral library from paired scan DDA (HCD + ETD) data, then quantifies DIA data using the library. Load both DDA and DIA data on the workflow tab and set the Data Type to "DIA-quant" for all DIA files.
<br>

##### isoDTB-ABPP
Perform closed search and MS1-based label quantification with chemically labelled cysteines. Details can be found from <i>Profiling the proteome-wide selectivity of diverse electrophiles</i> (10.33774/chemrxiv-2021-w7rss-v2). Note: If the dataset has multiple MS files (e.g. replicates or LC fractions) they should be annotated as different experiments (e.g. F1, F2, etc. in the Experiment column).
<br>

##### diaTOP_ABPP
Workflow for cysteine activity-based protein profiling method with DIA-based label-free quantification, based on data described in: "Quantitative Chemoproteomic Profiling with Data-Independent Acquisition-Based Mass Spectrometry", Fan Yang et al. JACS 144:901\u2013911 (2022). This workflow supports DDA-based (when DDA data is available), DIA-based (directDIA via MSFragger-DIA), and hybrid (from both DIA and DDA) spectral library building. MSFragger (MSFragger-DIA for DIA data) search with a static modification of +57.02146 on cysteine (iodoacetamide alkylation) specified as fixed modification, and +223.1685 is specified on cysteine to account for the probe mass. In addition, Met oxidation and N-terminal Acetyl are specified as variable modifications. 20 ppm precursor tolerance, parameter optimization enabled. Fully tryptic search, up to 2 missed cleavages. Deep learning-based rescoring of MSFragger results with MSBooster, followed by PSM validation with Percolator. Protein inference is performed with ProteinProphet. The data is filtered to 1% protein FDR and additionally at 1% FDR at the peptide and PSM levels. Spectral library building with EasyPQP followed by quantification, using that library, from the DIA files using DIA-NN.
<br>

##### ipIAA-ABPP
Workflow for the analysis of chemical proteomics data for profiling cysteines using isotopic isopropyl iodoacetamide alkyne probes (IPIAA). Closed search with MSFragger, PSM validation using Percolator, isotope labeling based (L\:H) quantification using IonQuant. Details in Heta et al. \u201CSP3-enabled rapid and high coverage chemoproteomic identification of cell-state dependent redox-sensitive cysteines\u201D, Mol Cell Proteomics 21(4), 100218. Note\: If the dataset has multiple MS files (e.g. replicates or LC fractions) they should be annotated as different experiments (e.g. F1, F2, etc. in the Experiment column).
<br>

##### isoTOP-ABPP
Workflow for cysteine activity-based protein profiling method isoTOP-ABPP: Weerapana, E., Wang, C., Simon, G.&nbsp;et al.&nbsp;Quantitative reactivity profiling predicts functional cysteines in proteomes.&nbsp;Nature&nbsp;468,&nbsp;790\u2013795 (2010). static modification of +57.02146 on cysteine (iodoacetamide alkylation) is specified as fixed modification, and +464.28596 (light probe) and +470.29977 (heavy probe) were specified on cysteine to account for probe modifications with the either light or heavy variants of the IA-probe-TEV adduct. In addition, Met oxidation and N-terminal Acetyl are specified as variable modifications. 20 ppm precursor tolerance, parameter optimization enabled. Fully tryptic search, up to 2 missed cleavages. Rescoring with MSBooster. Validation with Percolator and ProteinProphet. Relative quantification of peptides with light/heavy probe using IonQuant. Note: If the dataset has multiple MS files (e.g. replicates or LC fractions) they should be annotated as different experiments (e.g. F1, F2, etc. in the Experiment column).
<br>

##### SLC-ABPP
Workflow for streamlined cysteine activity-based protein profiling method (SLC-ABPP) based on Kuljanin et al. \u201CReimagining high-throughput profiling of reactive cysteines for cell-based screening of large electrophile libraries\u201D. Nat Biotechnol 39, 630\u2013641 (2021). TMT 16-plex, quantification from MS3 with real-time-search (RTS) option and identification form low mass accuracy MS2 (ion trap). Cysteines are labeled with DBIA (+239) which is specified, in addition to Met oxidation, as a variable modification. Cystein carbamethyl is specified as fixed modification. 20 ppm precursor tolerance (parameter optimization enabled). Fully tryptic, up to 2 missed cleavages. In scoring, only single charged fragments are used (changed for DBIA data from the default, which is maximum fragment charge of 2). Rescoring with MSBooster. Validation with Percolator and ProteinProphet. TMT-Integrator with virtual reference approach, median-centering normalization, data summarization at all levels (with C[399], i.e. Cys+DBIA, for the site-level reports; localization directly from MSFragger).
<br>

##### PAL
Workflow for photoaffinity labeling (PAL) data, mass offset search. Light probe mass: 436.22565, heavy probe mass: 442.2633. L:H quantification with IonQuant. For other probes, changes modification masses in MSFragger and IonQuant tabs accordingly. For PTM-Shepherd analysis of peptides with the mass shift, make the following changes: in MSFragger tab change 'Report mass shift as a variable mod' to No; In Quant (MS1 tab) tab switch from IonQuant to Philosopher freequant.
<br>

##### WWA
Perform database searching for wide-window acquisition (WWA) data, followed by label free quantification with match-between-runs using IonQuant. Need to set the "data type" to WWA.
<br>
<br>

#### Next: see the [FragPipe usage tutorial](https://fragpipe.nesvilab.org/docs/tutorial_fragpipe.html).
